tests/end-to-end/upgrade/internal/runner/runner.go

```diff
@@ -0,0 +1,247 @@
+package runner
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/kyma-project/kyma/tests/end-to-end/upgrade/internal/platform/logger"
+
+	"github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	gracefulTimeout = time.Second * 10
+	runnerLabelName = "upgrade.tester.kyma-project.io"
+	// negative value for regex used for name validation in k8s:
+	// https://github.com/kubernetes/apimachinery/blob/98853ca904e81a25e2000cae7f077dc30f81b85f/pkg/util/validation/validation.go#L110
+	regexSanitize = "[^a-z0-9]([^-a-z0-9]*[^a-z0-9])?"
+)
+
+// NamespaceCreator has methods required to create k8s ns.
+type NamespaceCreator interface {
+	Create(*v1.Namespace) (*v1.Namespace, error)
+}
+
+// TestRunner executes registered tests
+type TestRunner struct {
+	log                 *logrus.Entry
+	nsCreator           NamespaceCreator
+	tests               map[string]UpgradeTest
+	maxConcurrencyLevel int
+	sanitizeRegex       *regexp.Regexp
+}
+
+// NewTestRunner is a constructor for TestRunner
+func NewTestRunner(log *logrus.Entry, nsCreator NamespaceCreator, tests map[string]UpgradeTest, maxConcurrencyLevel int) (*TestRunner, error) {
+	sanitizeRegex, err := regexp.Compile(regexSanitize)
+	if err != nil {
+		return nil, errors.Wrap(err, "while compiling sanitize regexp")
+	}
+
+	return &TestRunner{
+		log:                 log.WithField("service", "test:runner"),
+		nsCreator:           nsCreator,
+		tests:               tests,
+		maxConcurrencyLevel: maxConcurrencyLevel,
+		sanitizeRegex:       sanitizeRegex,
+	}, nil
+}
+
+// PrepareData executes CreateResources method for each registered test.
+// Test are run in parallel with given maxConcurrencyLevel
+func (r *TestRunner) PrepareData(stopCh <-chan struct{}) error {
+	wg := sync.WaitGroup{}
+	queue := make(chan task)
+
+	failedTaskCnt := 0
+	// spawn workers
+	for i := 0; i < r.maxConcurrencyLevel; i++ {
+		wg.Add(1)
+		go func() {
+			for task := range queue {
```

Comment:
IMO the variable name should be different than the struct name used in PrepareData function, it could be confusing when you read it.